### PR TITLE
zipkin: Fix wrongly rendered timestamp value

### LINKED ISSUE #9341: Zipkin HTTP_JSON v2 spans using strings for timestamp/duration
*Title*: Zipkin HTTP_JSON v2 payload using strings for timestamp/duration

*Description*:
According to the Zipkin v2 API (https://zipkin.io/zipkin-api/#/default/post_spans) spans should contain `timestamp` and `duration` fields as `int64`, not strings, in the JSON payload. This makes collectors not being able to parse the JSON payload.

When configuring Zipkin to use the `HTTP_JSON` endpoint version, the resulting payload is:

```
[
  {
    "traceId": "83713138fa14c726",
    "id": "83713138fa14c726",
    "kind": "SERVER",
    "name": "www.hello",
    "timestamp": "1572849988488551",  <-------------------------------------------
    "duration": "7131",        <-------------------------------------------
    "localEndpoint": {
      "serviceName": "envoy-cluster",
      "ipv4": "172.31.0.5"
    },
    "tags": {
      "response_flags": "-",
      "request_size": "0",
      "response_size": "292",
      "node_id": "envoy",
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "testcluster",
      "guid:x-request-id": "3ab0fc56-f92e-9c03-a087-1f23662699cb",
      "zone": "us-east-1",
      "http.protocol": "HTTP/1.1",
      "component": "proxy",
      "http.url": "https://www.hello/headers",
      "downstream_cluster": "-",
      "user_agent": "HTTPie/0.9.2"
    }
  }
]
```
When it should be

```
[
  {
    "traceId": "83713138fa14c726",
    "id": "83713138fa14c726",
    "kind": "SERVER",
    "name": "www.hello",
    "timestamp": 1572849988488551,  <-------------------------------------------
    "duration": 7131,        <-------------------------------------------
    "localEndpoint": {
      "serviceName": "envoy-cluster",
      "ipv4": "172.31.0.5"
    },
    "tags": {
      "response_flags": "-",
      "request_size": "0",
      "response_size": "292",
      "node_id": "envoy",
      "http.status_code": "200",
      "http.method": "GET",
      "upstream_cluster": "testcluster",
      "guid:x-request-id": "3ab0fc56-f92e-9c03-a087-1f23662699cb",
      "zone": "us-east-1",
      "http.protocol": "HTTP/1.1",
      "component": "proxy",
      "http.url": "https://www.hello/headers",
      "downstream_cluster": "-",
      "user_agent": "HTTPie/0.9.2"
    }
  }
]
```

*Repro steps*:
Enable Zipkin tracing with `HTTP_JSON` as follows:

```
tracing:
  http:
    name: envoy.zipkin
    config:
      collector_cluster: tracing-sidecar
      collector_endpoint: /api/v2/spans
      collector_endpoint_version: HTTP_JSON
      shared_span_context: false
```

### PR DESCRIPTION
**Pull request by dio**
This patch fixes the wrongly rendered timestamp value of serialized Zipkin span by introducing replacers. This path is chosen since the way protobuf struct holds numeric data is by using double as its data type and when it is serialized (through `MessageToJsonString`; internally by `snprintf`-ing the value to a string) sometimes lead to a value with scientific notation involved (e.g. `1584324295476870` will be printed as `1.58432429547687e+15`, as reported in #9341). 

A better solution probably by going down to fix protobuf `MessageToJsonString`.

Risk Level: Low
Testing: Unit-tested
Docs Changes: N/A
Release Notes: N/A
Fixes #9341

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

**Comment by marcantoine-bibeau** [comment]
is your image base on the same as Envoy. Seems not cause "apt" is not available :( 
```
Sending build context to Docker daemon  3.733MB
Step 1/8 : FROM dio123/envoy-zipkin:9341-fix
9341-fix: Pulling from dio123/envoy-zipkin
c9b1b535fdd9: Already exists
11be6d113a7c: Pull complete
5c68e248835e: Pull complete
8f5719c7a9c2: Pull complete
4ab467168a5c: Pull complete
ca555b013c37: Pull complete
Digest: sha256:8a191866cafb8d9bfc51091fe1ebc1f167b8acb90200c4033c76cb442d172ea0
Status: Downloaded newer image for dio123/envoy-zipkin:9341-fix
 ---> e8a2ebe23b9a
Step 2/8 : RUN apt update && apt-get install -y     iproute2     iptables     gettext-base     curl && rm -rf /var/lib/apt/lists/*
 ---> Running in f37fb2520d45
/bin/sh: apt: not found
The command '/bin/sh -c apt update && apt-get install -y     iproute2     iptables     gettext-base     curl && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 127
```

**Comment by dio** [comment]
@marcantoine-bibeau sorry, that was alpine. I have re-pushed it. Could you help to re-pull it again?

**Comment by marcantoine-bibeau** [comment]
yep :) doing it now!

**Comment by marcantoine-bibeau** [comment]
ok image looks ok, running app with some tracing to see if http 400 is happening... So far so good but let me runs this for 1 hour... 

Let me get back to you in about 1 hour with the results!

**Comment by dio** [comment]
@marcantoine-bibeau awesome. Thank you!

**Comment by marcantoine-bibeau** [comment]
So after a bit more than 30min, no http 400 when sending zipkin v2 traces, ix seems perfect :) Thank you very very much. Will you be able to merge that back to 13.x?

**Comment by marcantoine-bibeau** [comment]
All Good :) Thanks again!

**Comment by junr03** [comment]
@dio you mention `A better solution probably by going down to fix protobuf MessageToJsonString.`. After reading the change I wonder if it is safer to go ahead and do the work necessary to ensure that we get the correct/desired value in the first place. It seems precarious to me to start holding state about replacements, and edit strings.

**Comment by dio** [comment]
@junr03 yes, the possible changes will be introducing control how we render double value* inside the protobuf's strutil https://github.com/protocolbuffers/protobuf/blob/422053f3bcb39cac483d2769e936c473e7c8bcdb/src/google/protobuf/stubs/strutil.cc#L1248-L1290, since rendering `uint64_t` will be string: https://github.com/protocolbuffers/protobuf/issues/2679 (well a flag probably can be introduced here as well). However, I think the "replacements" are pretty well-scoped and especially for the v2 serializer.

*double is chosen since we use struct as "container" before rendering it to (JSON) string using protobuf lib. We also do the same thing with the "typed" access log.

**Comment by junr03** [comment]
So what I am hearing is that while you'd like to do that long term, you want to proceed with this PR as a short term fix to solve #9341, right?

If that is the case, do you mind creating issues and linking inline so we can keep track of that? After that I'll go ahead and review this approach more closely. Does that sound good to you?

**Comment by dio** [comment]
I have logged here. https://github.com/envoyproxy/envoy/issues/10411. I also think that probably keeping it as string probably the safest way, but it needs an effort on documenting/enforcing it (e.g. by helping the collectors out there to embrace it).

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 1]
```diff
@@ -71,20 +73,23 @@ JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
...
-                               [](std::string* element, const ProtobufWkt::Struct& span) {
-                                 absl::StrAppend(element, MessageUtil::getJsonStringFromMessage(
-                                                              span, false, true));
-                               }));
+            out,
+            absl::StrJoin(toListOfSpans(zipkin_span, replacements), ",",
+                          [replacements](std::string* element, const ProtobufWkt::Struct& span) {
+                            const std::string json =
+                                MessageUtil::getJsonStringFromMessage(span, false, true);
+                            absl::StrAppend(element, absl::StrReplaceAll(json, replacements));
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]: inline comment about why we are replacing the strings with integers.

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 2]
```diff
@@ -176,7 +204,7 @@ TEST(ZipkinSpanBufferTest, SerializeSpan) {
                            R"("traceId":"0000000000000001",)"
                            R"("id":"0000000000000001",)"
                            R"("kind":"CLIENT",)"
-                           R"("timestamp":1566058071601051,)"
+                           R"("timestamp":DEFAULT_TEST_TIMESTAMP,)"
```
Comment (junr03) [inline test/extensions/tracers/zipkin/span_buffer_test.cc]: I might be being dense. I am not sure I understand how this expectation is met given that the DEFAULT_TEST_TIMESTAMP string is not being replaced?

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 3]
```diff
@@ -60,9 +61,10 @@ SerializerPtr SpanBuffer::makeSerializer(
 }
 
 std::string JsonV1Serializer::serialize(const std::vector<Span>& zipkin_spans) {
```
[the remark this answers is not in the document]
Comment (dio) [reply source/extensions/tracers/zipkin/span_buffer.cc]: Yes, it is deprecated. Not sure about the schedule though. As of today, it is available as one example of using deprecated config here: https://github.com/envoyproxy/envoy/blob/c6232b702e1bfa2185041883666a5add2965d9e1/configs/using_deprecated_config.v2.yaml#L70

File: test/extensions/tracers/zipkin/span_buffer_test.cc (hunk 2 above)
Comment (dio) [reply test/extensions/tracers/zipkin/span_buffer_test.cc]: Sorry for not being really clear about this. But the `array_string` param of  `wrapAsObject` will replace each `DEFAULT_TEST_TIMESTAMP` placeholder with `DEFAULT_TEST_TIMESTAMP` value above (by calling `withDefaultTimestamp` function), i.e. in this case: `1584324295476870`.

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 4]
```diff
@@ -103,7 +108,7 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span) const {
     }
 
     if (annotation.isSetEndpoint()) {
-      (*fields)[SPAN_TIMESTAMP] = ValueUtil::numberValue(annotation.timestamp());
+      (*fields)[SPAN_TIMESTAMP] = Util::uint64Value(annotation.timestamp(), replacements);
```
[the remark this answers is not in the document]
Comment (dio) [reply source/extensions/tracers/zipkin/span_buffer.cc]: Done. Not sure if the comment is clear enough.

**Comment by SecurityInsanity** [comment]
***tl;dr*** I think there's actual merit to merging this in for the short term, and chasing down a more permanent fix in protobuf (as mentioned in the comments of these PRs)

I totally understand the nervousness at doing the conversion at this layer (It's totally not the most ideal)!  However, at the same time it'll probably take awhile to get the fix upstream'd in protobuf (and a release cut, even when it is merged in). Since the code in protobuf is used for both actual doubles, and uint64_t's it has to be done with an extra branch, and some math trickery. Not to mention getting golang (or other languages with this problem) to parse scientific notation. Probably has more edge cases in the end, than us doing this here.

File: test/extensions/tracers/zipkin/span_buffer_test.cc (hunk 2 above)
Comment (junr03) [reply test/extensions/tracers/zipkin/span_buffer_test.cc]: Nope, that is my bad, I didn't connect the two 🤦‍♂

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 5]
```diff
@@ -71,20 +72,44 @@ JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
...
+                           span, /* pretty_print */ false,
+                           /* always_print_primitive_fields */ true);
+
+                       // The Zipkin API V2 specification mandates to store timestamp value as int64
+                       // https://github.com/openzipkin/zipkin-api/blob/228fabe660f1b5d1e28eac9df41f7d1deed4a1c2/zipkin2-api.yaml#L447-L463
+                       // (often translated as uint64 in some of the official implementations:
+                       // https://github.com/openzipkin/zipkin-go/blob/62dc8b26c05e0e8b88eb7536eff92498e65bbfc3/model/span.go#L114,
+                       // and see the discussion here:
+                       // https://github.com/openzipkin/zipkin-go/pull/161#issuecomment-598558072),
+                       // however when the timestamp is stored as number value (double) there is a
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]:
```suggestion
                       // https://github.com/openzipkin/zipkin-go/pull/161#issuecomment-598558072).
                       // However, when the timestamp is stored as number value in a protobuf struct, it is stored as a double. Because of how protobuf serializes doubles, there is a
```

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 6]
```diff
@@ -71,20 +72,44 @@ JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
...
+                       // possibility that the value will be rendered as a number with scientific
+                       // notation as reported in:
+                       // https://github.com/envoyproxy/envoy/issues/9341#issuecomment-566912973. To
+                       // deal with that issue, here we do a workaround by storing the timestamp as
+                       // string and keeping track that with the corresponding integer replacements,
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]:
```suggestion
                       // string and keeping track of that with the corresponding integer replacements,
```

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 7]
```diff
@@ -71,20 +72,44 @@ JsonV2Serializer::JsonV2Serializer(const bool shared_span_context)
...
+                       // and do the replacement here so we can meet the Zipkin API V2 requirements.
+                       //
+                       // TODO(dio): The right fix for this is to introduce additional knob when
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]: great, thanks for adding this!

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 8]
```diff
@@ -103,7 +128,14 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span) const {
     }
 
     if (annotation.isSetEndpoint()) {
-      (*fields)[SPAN_TIMESTAMP] = ValueUtil::numberValue(annotation.timestamp());
+      // Usually we store number to a ProtobufWkt::Struct object via ValueUtil::numberValue, however
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]:
```suggestion
      // Usually we store number to a ProtobufWkt::Struct object via ValueUtil::numberValue. However,
```

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 9]
```diff
@@ -103,7 +128,14 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span) const {
...
+      // due to the possibility of rendering that to a number with scientific notation, we chose to
+      // store it as a string and keeping track the corresponding replacement. e.g. we have
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]:
```suggestion
      // store it as a string and keep track the corresponding replacement. For example, we have
```

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 10]
```diff
@@ -103,7 +128,14 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span) const {
...
+      // 1584324295476870 if we stored it as a double value, MessageToJsonString gives
+      // us 1.58432429547687e+15, instead we store it as the string of 1584324295476870 (when it is
```
Comment (junr03) [inline source/extensions/tracers/zipkin/span_buffer.cc]:
```suggestion
      // us 1.58432429547687e+15. Instead we store it as the string of 1584324295476870 (when it is
```

File: source/extensions/tracers/zipkin/zipkin_core_types.h [hunk 11]
```diff
@@ -37,15 +38,19 @@ class ZipkinBase {
    * All classes defining Zipkin abstractions need to implement this method to convert
    * the corresponding abstraction to a ProtobufWkt::Struct.
    */
-  virtual const ProtobufWkt::Struct toStruct() const PURE;
+  virtual const ProtobufWkt::Struct toStruct(Util::Replacements& replacements) const PURE;
```
Comment (junr03) [inline source/extensions/tracers/zipkin/zipkin_core_types.h]: Add param comment about replacements.

Comment (dio) [reply source/extensions/tracers/zipkin/zipkin_core_types.h]: 👍

File: source/extensions/tracers/zipkin/util.cc [hunk 12]
```diff
@@ -22,6 +23,12 @@ uint64_t Util::generateRandom64(TimeSource& time_source) {
   return rand_64();
 }
 
+ProtobufWkt::Value Util::uint64Value(uint64_t value, Replacements& replacements) {
+  const std::string string_value = std::to_string(value);
+  replacements.push_back({absl::StrCat("\"", string_value, "\""), string_value});
```
Comment (htuch) [inline source/extensions/tracers/zipkin/util.cc]: I'm assuming there is no possibility that an integer that is called for replacement in a replacement context conflicts with an unrelated integer which happens to share the value? i.e. we have a replacement for 42, and accidentally substitute an unrelated 42?

**Comment by dio** [comment]
I can leave #9341 open (and add a specific issue for that). @htuch, I think TODO is added.

File: source/extensions/tracers/zipkin/util.cc (hunk 12 above)
Comment (dio) [reply source/extensions/tracers/zipkin/util.cc]: I think I'll add a precautionary by adding the field name as well.

**Comment by dio** [comment]
Will add test as per @adriancole.

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 13]
```diff
@@ -265,102 +303,150 @@ TEST(ZipkinSpanBufferTest, SerializeSpan) {
...
+TEST(ZipkinSpanBufferTest, TestSerializeTimestampInTheFuture) {
+  ProtobufWkt::Struct objectWithScientificNotation;
+  auto* objectWithScientificNotationFields = objectWithScientificNotation.mutable_fields();
+  (*objectWithScientificNotationFields)["timestamp"] = ValueUtil::numberValue(
+      DEFAULT_TEST_TIMESTAMP); // the value of DEFAULT_TEST_TIMESTAMP is 1584324295476870.
+  const auto objectWithScientificNotationJson =
+      MessageUtil::getJsonStringFromMessage(objectWithScientificNotation, false, true);
+  // Since we use ValueUtil::numberValue to set the timestamp, we expect to
+  // see the value is rendered with scientific notation (1.58432429547687e+15).
+  EXPECT_EQ(R"({"timestamp":1.58432429547687e+15})", objectWithScientificNotationJson);
```
Comment (dio) [inline test/extensions/tracers/zipkin/span_buffer_test.cc]: Here we have `1.58432429547687e+15` (the input is `1584324295476870`) if we set the field using `ValueUtil::numberValue`.

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 14]
```diff
@@ -265,102 +303,150 @@ TEST(ZipkinSpanBufferTest, SerializeSpan) {
...
+
+  ProtobufWkt::Struct object;
+  auto* objectFields = object.mutable_fields();
+  Util::Replacements replacements;
+  (*objectFields)["timestamp"] = Util::uint64Value(DEFAULT_TEST_TIMESTAMP, replacements);
+  const auto objectJson = MessageUtil::getJsonStringFromMessage(object, false, true);
+  // We still have "1584324295476870" from MessageUtil::getJsonStringFromMessage here.
+  EXPECT_EQ(R"({"timestamp":"1584324295476870"})", objectJson);
```
Comment (dio) [inline test/extensions/tracers/zipkin/span_buffer_test.cc]: Before replacement, `Util::uint64Value` stores the value as a string. `"1584324295476870"`.

File: test/extensions/tracers/zipkin/span_buffer_test.cc [hunk 15]
```diff
@@ -265,102 +303,150 @@ TEST(ZipkinSpanBufferTest, SerializeSpan) {
...
+  // However, then the replacement correctly replaces "1584324295476870" with 1584324295476870
+  // (without quotes).
+  EXPECT_EQ(R"({"timestamp":1584324295476870})", absl::StrReplaceAll(objectJson, replacements));
```
Comment (dio) [inline test/extensions/tracers/zipkin/span_buffer_test.cc]: Here we remove the `"` via replacement.

**Comment by dio** [comment]
@junr03 do you have more comments on this? Thanks!

**Comment by mattklein123** [comment]
Can you merge master and we can get this merged?

/wait

**Comment by objectiser** [comment]
@dio @mattklein123 Seems like this fix is having a side effect - the string replacements are causing other (tag) fields with the same numeric value as the duration, to have their quotes stripped. See https://github.com/istio/istio/issues/24177.

File: source/extensions/tracers/zipkin/util.cc (hunk 12 above)
Comment (douglas-reid) [reply source/extensions/tracers/zipkin/util.cc:28]: This actually happened. We need a follow-on PR that adds the field name.

Comment (douglas-reid) [reply source/extensions/tracers/zipkin/util.cc:28]: See: https://github.com/envoyproxy/envoy/pull/11374

**Comment by dio** [comment]
@objectiser could you help to review @douglas-reid's PR as well? I've made couple of comments too. Thank you!

**Comment by owais** [comment]
Thanks for fixing this @dio. Do you know if this has made into any releases yet?

**Comment by moderation** [comment]
Doesn't look like the merge commit is in a release yet. You would have to user 1.15.0-dev / master for this fix.

`git tag --contains 01fe4b620f643bea06bdf5cea3443f9b0767fc40`
